### DISCUSSION TITLE
Unreal Engine Horde

### DISCUSSION THREAD

**Discussion by ghost** [aws-games/cloud-game-development-toolkit#217] [Ideas]
The design and implementation of a Terraform module for provisioning Unreal Engine Horde build system on AWS is in the works. We'd like to get feedback from the public on features this module should have. This is also a great place to ask general questions concerning Horde deployments.

**Comment by pasotee** [comment]
Hey @henrykie, it's exciting to see AWS aware of this. We plan to deploy Horde via Terraform by the end of this month. 

Is there something WIP you could share? I am super curious to see what your approach to architecture will be. We are currently building around ECS and the Docker images via Unreal's Build Graph.

**Reply by pasotee** [reply]
I am sorry, I am blind. I saw your submit from earlier today. I will get digging through it.

**Reply by ghost** [reply]
No worries! Happy to chat if you need.

The module currently pulls the docker containers from Epic's GHCR. You can also build your own, but will need to update the module to pull the image from somewhere else. This is configured here: https://github.com/aws-games/cloud-game-development-toolkit/blob/c2589d997c2af3f2f47c529d39275b1d35ec35c3/modules/unreal/horde/local.tf#L11

**Reply by pasotee** [reply]
Thanks for the reply. We've been playing around with your module since and spotted it the source.

We had some troubles with the DNS setup, but that's all on us and mainly caused by our domain registrar.
We will most likely end up forking it and change the URL/parameterising the image you highlighted it but still build on top of it.

Thanks again for going through the trouble of building it.

**Comment by cw255** [comment]
Nice to have: include EFS volume in terraform to store P4TRUST for globals.json (or other solution)

i.e.
Horde__configPath=//depot/globals.json
Horde__perforce__0__id=default
Horde__perforce__0__serverAndPort=ssl:server:1666
Horde__perforce__0__credentials__userName=
Horde__perforce__0__credentials__password=
P4TRUST=/efs-volume/.p4trust

where /efs-volume/.p4trust is pre-populated with fingerprint of ssl:server:1666

I don't think there is any native way to tell Horde to trust P4SERVER since I believe Epic don't utilise SSL internally. There was a feature request with Epic UE-250538 to do this natively in Horde environment but I don't know if its possible to track that and I don't see it in current docs

**Comment by medeveal** [comment]
Greetings @henrykie !

We've run into some issues using Horde deployed to ECS 
`Error starting gRPC call. 
HttpRequestException: The HTTP/2 server reset the stream. HTTP/2 error code 'PROTOCOL_ERROR' (0x1). (HttpProtocolError) .`


Maybe it would be a nice addition to have a possibility to switch from alb to nlb for grpc traffic in this case ? 
or extend rules of lb routing if Horde Agents are not configured to sent strict headers ? 
`"HTTP header (value) content-type = application/grpc"`

**Reply by ghost** [reply]
Hello! Thank you for reaching out. We route traffic based on headers because both the agent and web communication requests arrive on the same port. You could extend the module to define two different DNS endpoints that resolve to distinct load balancers pointing at the distinct server target groups used for web traffic and GRPC.

I could add the target groups for the web and grpc traffic as outputs so that custom deployments with separate load balancers would be easier. Would that be helpful?
